File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        int n=nums.size();
        backtrackfunction(nums,result,0,n-1);
        return result;
        
                }
        
        
        
       void  backtrackfunction(vector<int>& nums, vector<vector<int>> &result,int start,int end)
        {
            if(start==end)
                result.push_back(nums);
            
          for(int i=start;i<=end;i++)
          {
              swap(nums[i],nums[start]);
              backtrackfunction(nums,result,start+1,end);
              swap(nums[start],nums[i]);
          }
            
            
            
            
        }
        
    
};

File: 0046-permutations/0046-permutations.py
class Solution:
    def backtrack(self, nums, result, temp):
        if len(temp) == len(nums):
            result.append(temp[:])
            return 
        for i in range(len(nums)):
            if nums[i] in temp:
                continue
            
            temp.append(nums[i])
            self.backtrack(nums, result, temp)
            temp.pop()

    def permute(self, nums):
        self.result = []
        self.temp = []
        self.backtrack(nums, self.result, self.temp)
        return self.result
